src/enkf/gridio_fv3reg.f90

```diff
@@ -1128,75 +1123,1383 @@ subroutine readgriddata_pnc(vars3d,vars2d,n3d,n2d,levels,ndim,ntimes, &
+            do k=1,nlevs
+               call mpi_gatherv(locworkvar3d(1:nxloc,1:nyloc,k), recvcounts2d(iope+1), mpi_real4, workvar3d(:,:,k), recvcounts2d, displs2d,&
+                        mpi_real4, 0,iocomms(mem_pe(nproc)) ,iret)
+            enddo
+            if(iope == 0 ) then
+              do k=1,nlevs
+                  nn = nn_tile0
+              do j=1,ny_res
+                 do i=1,nx_res
+                    nn=nn+1
+                    workinc3d(i,j,k)=vargrid(nn,levels(qnr_ind-1)+k,nb,ne)
+                 enddo
+              enddo
+              enddo
+              workvar3d=workvar3d+workinc3d
+               if ( l_use_enkf_directZDA .and. cliptracers ) then ! set cliptracers to remove negative hydrometers
+                   clip = tiny(workvar3d(1,1,1))
+                   where (workvar3d < clip) workvar3d = clip
+               end if
+            endif
+            do k=1,nlevs
+               call mpi_scatterv(workvar3d(:,:,k),recvcounts2d,displs2d,mpi_real4,&
+                     locworkvar3d(:,:,k),recvcounts2d(iope+1),mpi_real4,0,iocomms(mem_pe(nproc)),ierror)
+            enddo
+            call write_fv3_restart_data3d(varstrname,fv3filename,file_id,locworkvar3d)
+
+         endif
+         if (ps_ind > 0) then
+           if(iope == 0)   allocate(workprsi(nx_res,ny_res,nlevsp1))
+           if(iope == 0)   allocate(pswork(nx_res,ny_res))
+           varstrname = 'delp'
+           call fv3lamfile%get_idfn(varstrname,file_id,fv3filename)
+           call read_fv3_restart_data3d(varstrname,filename,file_id,locworkvar3d)   ! Pascal
+           do k=1,nlevs
+              call mpi_gatherv(locworkvar3d(1:nxloc,1:nyloc,k), recvcounts2d(iope+1), mpi_real4, workvar3d(:,:,k), recvcounts2d, displs2d,&
+                        mpi_real4, 0,iocomms(mem_pe(nproc)) ,iret)
+           enddo
+           if(iope == 0 ) then
+              workprsi(:,:,nlevsp1)=eta1_ll(nlevsp1) !etal_ll is needed
+              do i=nlevs,1,-1
+                 workprsi(:,:,i)=workvar3d(:,:,i)*0.01_r_kind+workprsi(:,:,i+1)
+              enddo
+
+              nn = nn_tile0
+              do j=1,ny_res
+              do i=1,nx_res
+                 nn=nn+1
+                 pswork(i,j)=vargrid(nn,levels(n3d)+ps_ind,nb,ne)
+              enddo
+              enddo
+              if(l_pres_add_saved) then
+                do k=1,nlevs+1
+                do j=1,ny_res
+                 do i=1,nx_res
+                    workinc3d2(i,j,k)=eta2_ll(k)*pswork(i,j)
+                 enddo
+                enddo
+                enddo
+                workprsi=workprsi+workinc3d2
+               else
+                  workprsi(:,:,1)=workprsi(:,:,1)+pswork
+                  do k=2,nlevsp1
+                     workprsi(:,:,k)=eta1_ll(k)+eta2_ll(k)*workprsi(:,:,1)
+                  enddo
+               endif
+               do k=1,nlevs
+                workvar3d(:,:,k)=(workprsi(:,:,k)-workprsi(:,:,k+1))*100.0
```

Comment:
Please add proper precision to real value.

Reply:
done